### PR TITLE
Adding command for New Database Connection

### PR DESCRIPTION
**Pull request by jhorvath**
Adding command for New Database Connection to java.lsp.server

### CODE REVIEW

File: java/java.lsp.server/nbproject/project.xml [hunk 1]
```diff
@@ -126,7 +132,15 @@
                     <compile-dependency/>
                     <run-dependency>
                         <release-version>1</release-version>
-                        <specification-version>1.82</specification-version>
+                        <specification-version>1.81</specification-version>
```
Comment (sdedic) [inline java/java.lsp.server/nbproject/project.xml:135]: Old version, was `1.82`. In fact this PR adds an API to db.explorer, so should even increase + apichanges & etc.

File: ide/db/src/org/netbeans/api/db/explorer/ConnectionManager.java [hunk 2]
```diff
@@ -187,6 +189,19 @@ public void removeConnection(DatabaseConnection dbconn) throws DatabaseException
...
         ConnectionList.getDefault().remove(dbconn.getDelegate());
     }
+    
+    /**
+     * Retrieves a list of database schemas available for given connection.
+     * 
+     * @param dbconn
+     * @return List of schemas or null if connection failed
+     */
+    public List<String> getSchemas(DatabaseConnection dbconn) {
```
Comment (JaroslavTulach) [inline ide/db/src/org/netbeans/api/db/explorer/ConnectionManager.java:199]: This is an API change. It needs `@since` tag, it needs a note in `apichanges.xml`.

File: ide/db/src/org/netbeans/api/db/explorer/ConnectionManager.java [hunk 3]
```diff
@@ -187,6 +189,19 @@ public void removeConnection(DatabaseConnection dbconn) throws DatabaseException
...
+        if (SwingUtilities.isEventDispatchThread()) {
+            throw new IllegalStateException("This method can not be called on the event dispatch thread."); // NOI18N
```
Comment (JaroslavTulach) [inline ide/db/src/org/netbeans/api/db/explorer/ConnectionManager.java:201]: _cannot_ is a single word - maybe you want to spell this (unusual?) restriction in the Javadoc of the method.

File: ide/db/src/org/netbeans/api/db/explorer/ConnectionManager.java [hunk 4]
```diff
@@ -187,6 +189,19 @@ public void removeConnection(DatabaseConnection dbconn) throws DatabaseException
 
         ConnectionList.getDefault().remove(dbconn.getDelegate());
     }
+    
+    /**
+     * Retrieves a list of database schemas available for given connection.
+     * 
+     * @param dbconn
+     * @return List of schemas or null if connection failed
```
Comment (JaroslavTulach) [inline ide/db/src/org/netbeans/api/db/explorer/ConnectionManager.java:197]: I usually use `{@code null}` to print it in a different font. Shouldn't an exception be raised if the connection fails?

File: ide/db/src/org/netbeans/modules/db/explorer/DbSchemasQuery.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                        while (rs.next()) {
+                            if (schemas == null) {
+                                schemas = new ArrayList<>();
+                            }
+                            schemas.add(rs.getString(1).trim());
+                        }
+                    }
+                }
+            }
+        } catch (DatabaseException | SQLException ex) {
```
Comment (JaroslavTulach) [inline ide/db/src/org/netbeans/modules/db/explorer/DbSchemasQuery.java:67]: Why not propagate the exception to the caller?

File: java/java.lsp.server/nbcode/nbproject/platform.properties [hunk 6]
```diff
@@ -30,7 +30,6 @@ cluster.path=\
 disabled.modules=\
```
Comment (JaroslavTulach) [inline java/java.lsp.server/nbcode/nbproject/platform.properties:30]: So many changes! However most of them are just reorderings, right? Hard to guess what was the real intent, can you revert the changes and start once again and create a minimal diff?

File: java/java.lsp.server/nbcode/nbproject/platform.properties [hunk 7]
```diff
@@ -310,153 +373,90 @@ disabled.modules=\
...
+    org.netbeans.modules.xsl,\
+    org.netbeans.spi.editor.hints.projects,\
+    org.netbeans.swing.dirchooser,\
+    org.netbeans.upgrader,\
+    org.openide.compat,\
+    org.openide.execution.compat8,\
+    org.openide.options,\
+    org.openide.util.enumerations,\
+    org.openidex.util
+nbjdk.active=JDK_11
```
Comment (JaroslavTulach) [inline java/java.lsp.server/nbcode/nbproject/platform.properties:458]: Avoid `nbjdk.active` and `nbplatform.active`.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/Server.java [hunk 8]
```diff
@@ -858,6 +859,7 @@ protected LanguageClient client() {
     public static final String JAVA_SUPER_IMPLEMENTATION =  "java.super.implementation";
     public static final String JAVA_SOURCE_FOR =  "java.source.for";
     public static final String GRAALVM_PAUSE_SCRIPT =  "graalvm.pause.script";
+    public static final String DB_ADD_CONNECTION =  "micronaut.add.db.connection";
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/Server.java:862]: Why is this command prefixed by `micronaut`? The DB explorer is completely micronaut unrelated, right?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ShowInputBoxParams.java [hunk 9]
```diff
@@ -83,6 +93,22 @@ public String getValue() {
...
+    @NonNull
+    public boolean isPassword() {
+        return password;
+    }
+
+    /**
+     * Controls if a password input is shown. Password input hides the typed text.
+     */
+    public void setPassword(boolean password) {
+        this.password = password;
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ShowInputBoxParams.java:110]: Expanding the NetBeans LSP protocol extensions. Brave!

Ideally we should version the changes, but it is a gray area so far. Consider at least increasing spec version of the `java.lsp.server` module and leaving a `@since` note on the new methods/field.

File: java/java.lsp.server/vscode/package.json [hunk 10]
```diff
@@ -401,6 +401,11 @@
 				"command": "foundProjects.deleteEntry",
 				"title": "Delete"
 			},
+      {
+                                "command": "micronaut.add.db.connection",
+                                "title": "Add Database Connection",
+ 				"category": "Micronaut"
```
Comment (JaroslavTulach) [inline java/java.lsp.server/vscode/package.json:407]: It'd be better if Micronaut extension were in `enterprise/micronaut` module, not here.

File: java/java.lsp.server/vscode/src/nbcode.ts [hunk 11]
```diff
@@ -97,7 +97,9 @@ if (typeof process === 'object' && process.argv0 === 'node') {
     let clusters = fs.readdirSync(nbcode).filter(c => c !== 'bin' && c !== 'etc').map(c => path.join(nbcode, c));
     let args = process.argv.slice(2);
     let json = JSON.parse("" + fs.readFileSync(path.join(extension, 'package.json')));
-    let globalStorage = path.join(os.homedir(), '.config', 'Code', 'User', 'globalStorage', json.publisher + '.' + json.name);
+    let globalStorage = os.platform() === 'darwin' ? 
+        path.join(os.homedir(), 'Library', 'Application Support', 'Code', 'User', 'globalStorage', json.publisher + '.' + json.name) :
```
Comment (JaroslavTulach) [inline java/java.lsp.server/vscode/src/nbcode.ts:101]: Thanks! Glad to have Mac developer who cares about debugging support!

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
Fine at first round:
- properly version the DB module API changes
- avoid needless changes in `platform.properties`
- write a test for the new method
- try to avoid references to "Micronaut" in this generic extension

File: ide/db/src/org/netbeans/modules/db/explorer/DbSchemasQuery.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+ */
+public class DbSchemasQuery {
+
+    /**
+     * Retrieves a list of database schemas available for given connection.
+     * 
+     * @param dbconn
+     * @return List of schemas or null if connection failed
+     */
+    public static List<String> getSchemasForConnection(org.netbeans.api.db.explorer.DatabaseConnection dbconn) {
```
Comment (matthiasblaesing) [inline ide/db/src/org/netbeans/modules/db/explorer/DbSchemasQuery.java:41]: Please use the metadata API. See `org.netbeans.modules.db.explorer.node.ConnectionNode.updateModel()` to see how the metamodel can be accessed and `org.netbeans.modules.db.explorer.node.SchemaNodeProvider.initialize()` for a usage sample.

File: ide/db/src/org/netbeans/api/db/explorer/ConnectionManager.java (hunk 2 above)
Comment (matthiasblaesing) [inline ide/db/src/org/netbeans/api/db/explorer/ConnectionManager.java:199]: There is no reason to add a new API at this level. See my other comment about the metadata api.

**Review by matthiasblaesing — CHANGES_REQUESTED** [review]

File: java/java.lsp.server/nbcode/nbproject/platform.properties (hunk 6 above)
Comment (JaroslavTulach) [reply java/java.lsp.server/nbcode/nbproject/platform.properties:30]: OK, now it is better. Adding:
```
org.netbeans.modules.db.core
org.netbeans.modules.db.dataview
```
seems sane. Not sure why we need `com.google.guava`!?

File: java/java.lsp.server/nbcode/nbproject/platform.properties [hunk 13]
```diff
@@ -30,7 +30,6 @@ cluster.path=\
 disabled.modules=\
     bcpg,\
     bcprov,\
-    com.google.guava,\
```
Comment (JaroslavTulach) [inline java/java.lsp.server/nbcode/nbproject/platform.properties:33]: Who needs Guava? I don't think NetBeans code is using it directly...

File: java/java.lsp.server/vscode/package.json [hunk 14]
```diff
@@ -401,6 +401,11 @@
 				"command": "foundProjects.deleteEntry",
 				"title": "Delete"
 			},
+                        {
+                                "command": "db.add.connection",
+                                "title": "Add Database Connection",
+ 				"category": "Database"
```
Comment (JaroslavTulach) [inline java/java.lsp.server/vscode/package.json:407]: Category "Database" sounds better.

**Review by JaroslavTulach — APPROVED** [review]
The most recent version eliminates the API changes altogether, great! The "Add Database Connection" command looks fine to me. Please squash a bit to eliminate the add/remove of `platform.properties` entries and the API attempt from the history after merge.

File: java/java.lsp.server/nbcode/nbproject/platform.properties (hunk 13 above)
Comment (jhorvath) [reply java/java.lsp.server/nbcode/nbproject/platform.properties:33]: It was added by mistake, it is not needed, removing

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+                                client.showInputBox(new ShowInputBoxParams(
+                                        Bundle.MSG_EnterPassword(), "", true)).thenAccept((password) -> { //NOI18N
+                                    if (password == null) {
+                                        return;
+                                    }
+                                    DatabaseConnection dbconn = DatabaseConnection.create(driver, u, username, null, password, true);
+                                    try {
+                                        ConnectionManager.getDefault().addConnection(dbconn);
+                                        List<String> schemas = getSchemas(dbconn);
+                                        System.out.println("schemas " + schemas);
```
Comment (matthiasblaesing) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java:102]: This should be a logger (if needed)

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+                                            client.showMessage(new MessageParams(MessageType.Info, Bundle.MSG_ConnectionAdded()));
+                                        } else {
+                                            List<QuickPickItem> schemaItems = schemas.stream().map(schema -> new QuickPickItem(schema)).collect(Collectors.toList());
+                                            client.showQuickPick(new ShowQuickPickParams(Bundle.MSG_SelectSchema(), schemaItems)).thenAccept((s) -> {
+                                                if (s == null) {
+                                                    try {
+                                                        // Command was interrupted, we don't want the connection to be added
+                                                        ConnectionManager.getDefault().removeConnection(dbconn);
+                                                    } catch (DatabaseException ex) {
+                                                        System.out.println(ex.getMessage());
```
Comment (matthiasblaesing) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java:113]: This should either be a logger or a call to `org.openide.util.Exceptions#printStackTrace`.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+            }
+        }
+
+        return client.showQuickPick(new ShowQuickPickParams(Bundle.MSG_SelectDriver(), false, items))
+                .thenApply(selectedItems -> {
+                    if (selectedItems == null) {
+                        return null;
+                    }
+                    if (!selectedItems.isEmpty()) {
+                        int i = ((Double) selectedItems.get(0).getUserData()).intValue();
```
Comment (matthiasblaesing) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java:81]: What am I missing here? Why is this a `Double` and not an `Integer` as declared in line 69? If this is a lossy conversion, I would cast to `Number`, shielding you from implementation specific brokenness.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+                                return;
+                            }
+                            client.showInputBox(new ShowInputBoxParams(
+                                    Bundle.MSG_EnterUsername(), "")).thenAccept((username) -> { //NOI18N
+                                if (username == null) {
+                                    return;
+                                }
+                                client.showInputBox(new ShowInputBoxParams(
+                                        Bundle.MSG_EnterPassword(), "", true)).thenAccept((password) -> { //NOI18N
+                                    if (password == null) {
```
Comment (matthiasblaesing) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java:95]: Is this intentional, that you open 3 consecutive prompts for the connection parameters? I think the current approach on the Swing side useing a single dialog is more user friendly (and the test connection button is a bonus :-))

Comment (sdedic) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java:95]: Note: this is an LSP-specific command implementation -- NetBeans have its standard Swing dialogs implemented in the DB module. Sadly LSP protocol does not allow (yet ?) for rich dialogs, so we need to approximate, I guess. 

Some DB - related commands (i.e. connect, disconnect) were bridged directly to NB implementations although that requried some fixes in NB codebase. So LSP module contains minimal stuff needed - as much as it can be done is done directly in IDE codebase.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java (hunk 17 above)
Comment (sdedic) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java:81]: The trouble is that the `QuickPickItem` structure is mashalled to the LSP connection stream and goes to the client and back. And because the client is (yuck!) javascript, it represents integer as `number` which is unmarshalled as double... At this moment I do not know if the culprit is eclipse lsp4j, it is a feature by design (who the knows what type 2 actually is in a JSON document) or the javascript existence itself. If it is Eclipse stuff, we could eventually file a bug against the library.
To improve possible future compatibility, I'd put there 
`((Number) selectedItems.get(0).getUserData()).intValue();`
so if ever numbers w/o floating part deserialize to Integers or Longs the code survives.

Comment (matthiasblaesing) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java:82]: That makes sense. Thanks for the pointer.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java (hunk 18 above)
Comment (matthiasblaesing) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java:96]: Yeah, I assumed that much - thanks for the confirmation.

**Review by matthiasblaesing — APPROVED** [review]
Thanks for the updates, looks sane to me.

**Comment by matthiasblaesing** [comment]
@jhorvath  Before this is merged:
- please ensure the author information is consistent, there is at least one merge commit, that only holds your username and a different email adress. I assume, this is created by github.
- after final review, please rebase onto master and squash into one commit (this will most likely fix the author point)
- I assume you read the license statement in the files you edited/added and you are in a position to donate the changes to the Apache Foundation and do so. (or you have an ICLA on file?)

**Comment by jhorvath** [comment]
> I have read and agree to the license statement.

File: java/java.lsp.server/nbcode/nbproject/platform.properties [hunk 19]
```diff
@@ -116,8 +116,6 @@ disabled.modules=\
     org.netbeans.modules.css.model,\
     org.netbeans.modules.css.prep,\
     org.netbeans.modules.css.visual,\
-    org.netbeans.modules.db.core,\
```
Comment (sdedic) [inline java/java.lsp.server/nbcode/nbproject/platform.properties:119]: Please check if those modules are actually needed - I couldn't find a reference to `db.core` or `db.dataview` APIs.

**Review by sdedic — CHANGES_REQUESTED** [review]

**Review by sdedic — APPROVED** [review]

**Comment by sdedic** [comment]
I would ignore the commit validation on Mac. Failed on a timeout.
